docs/proposals/mempool/mempool.md

```diff
@@ -0,0 +1,221 @@
+	// OR
+	// replaces an existing transaction
+	// sets executableIndex based on nonce
+}
+
+func(m *Mempool) getExecutableTxs(txType txtype.TransactionType) []models.GenericTransaction {
+	result := make([]models.GenericTransaction)
+	for userTx range userTxsMap {
+		if executableIndex == -1 {
+			continue
+		}
+		executableTx := userTx.txs[userTx.executableIndex]
+		if (executableTx.Type == txType) {
+			result = append(result, executableTx)
+		}
+	}
+	return result
+}
+
+func(m *Mempool) getNextExecutableTx(stateID uint32) models.GenericTransaction {
+	// checks if tx from userTxsMap for given user is executable, if so increments executableIndex by 1
+	// returns txs[executableIndex]
+}
+
+func(m *Mempool) ignoreUserTxs(stateID uint32) {
+	// makes subsequent getExecutableTxs not return transactions from this user state
+	// this virtually marks all user's txs as non-executable
+	m.userTxsMap[stateID].executableIndex = -1
+}
+
+func(m *Mempool) resetExecutableIndices() {
+	// iterate over all UserTxs and set executableIndex to 0
+}
+
+func(m *Mempool) removeTxsAndRebalance(txs []models.GenericTransaction) {
+	// remove given txs from the mempool and possibly rebalance txs list
+}
+
+func (m *Mempool) getExecutableIndex(stateID uint32) int {
+    // returns current executableIndex for given user 
+	return m.userTxsMap[stateID].executableIndex
+}
+
+func (m *Mempool) updateExecutableIndicesAndNonces(newExecutableIndicesMap map[uint32]int) {
+    for stateID, index := range newExecutableIndicesMap {
+		// calculate applied txs count and decrease nonce based on executableIndex difference
+        userTxs := m.userTxsMap[stateID]
+		txsCountDifference = userTxs.executableIndex - index
+        userTxs.executableIndex = index
+        userTxs.nonce -= txsCountDifference
+    }
+}
+```
+
+```go
+// Max heap sorted by tx fee
+type Heap struct {
+	// heap internals
+}
+
+func NewHeap(txs []models.GenericTransaction) *Heap {
+	// init the heap
+}
+
+func(h *Heap) peek() models.GenericTransaction {
+	// get reference to the tx with the highest fee
+}
+
+func(h *Heap) pop() models.GenericTransaction {
+	// remove the max tx from the heap
+}
+
+func(h *Heap) replace(tx models.GenericTransaction) {
+	// pops max element, pushes new one to the heap and rebalances
+}
+```
+
+```go
+type RollupContext struct {
+	mempool Mempool
+	sucessfulTxs []models.GenericTransaction
+	failedTxs []models.GenericTransaction
+	executableIndicesCache map[uint32]int // map StateID -> ExecutableIndex before each commitment
+}
+
+func (r *RollupContext) rollupLoopIteration(txType txtype.TransactionType) {
+	r.mempool.fetchIncomingTxs()
+	txs := r.mempool.getExecutableTxs(txType)
+	heap := NewHeap(txs)
+
+	
+	// Apply txs modifying Mempool and Heap data structures
+	commitments := make([]models.Commitment, 0, MAX_COMMITMENT)
+    for i := 0; i != MAX_COMMITMENT; i++{
+        r.executableIndicesCache = make(map[uint32]int, MAX_TXS_PER_COMMITMENT)
```

Comment:
I'm thinking about a slightly different design: Treat executableIndices as ephmeral data that is cleaned after each batch. This way we dont need to keep a cache of them and restore them after failed commitments. The flow will be as follows:

1. Starting to create new batch. Generate initial heap. This also increments executable indicess by one since first transactions are now in the heap. Initially they were at the beginning of the array.
2. Consumed transactions from the heap replaced using the transactions from the mempool. This also advances the executable index by one.
3. Consumed transactions are also tracked in `sucessfulTxs` and `failedTxs` lists.
4. Using the transactions from the lists mempool is cleaned and the nonces are advanced forward. Executable indicess should be reset back to the begginning of the array (or -1 for non-executable).

Reply:
In this case we can even store the `executableIndices` externally in the rollup loop iteration for optimisation/simplification purpoces